html: Remove dependency on phf

zecakeh opened the pull request:
Allows to put some entries behind a cargo feature, instead of having to duplicate the list for each feature.

The APIs are really different so some significant refactoring was necessary, but in my opinion it makes the spec rules definitions easier to read.

jplatte reviewed (commented):
If all we need is associate data with static strings, no iteration on the maps or such, could we replace these set and map macros with `match` expressions, and not use any dependencies?

zecakeh commented:
I was actually thinking the same thing :sweat_smile:

zecakeh committed:
html: Remove dependency on phf

Allows to put some entries behind a cargo feature, instead of having to
duplicate the list for each feature.

zecakeh commented on crates/ruma-html/src/sanitizer_config/clean.rs:
```diff
@@ -475,3 +383,173 @@ enum AttributeAction {
+    /// according to the Matrix specification.
+    pub(super) fn is_attribute_allowed(element_name: &str, attribute_name: &str) -> bool {
+        match element_name {
+            "span" => match attribute_name {
+                "data-mx-bg-color" | "data-mx-color" | "data-mx-spoiler" | "data-mx-maths" => true,
+                #[cfg(feature = "unstable-msc4286")]
+                "data-msc4286-external-payment-details" => true,
+                _ => false,
+            },
```
Sadly we can't use `matches!` for this because it doesn't support `cfg` attributes either.
